### DISCUSSION TITLE
Streaming Responses in a UI

### DISCUSSION THREAD

**Discussion by StephenDWright** [imartinez/privateGPT#684] [Q&A]
This may be a bit niche but has anyone who has tried a UI or created their own, figured out how to stream responses to the UI as it is generated. If so, using which library, e.g gradio/streamlit.. documentation for OpenAI is everywhere but limited information using llama models. 

Any success stories would be appreciated.

**Comment by sime2408** [comment]
Streamlit reruns the script from top to bottom for each interaction, so it's kind of tricky. I am searching for something similar. For now, I can come up with a lol solution for this 😆 :

```python

# wait for the answer
st.write("Bot is typing...")
answer, source_documents = query_documents(user_input, selected_database, selected_collection)
message(answer, key=answer_key, avatar_style="bottts", seed=5)
```
I also used `with st.spinner("Processing..."):`

see: [my fork](https://github.com/sime2408/scrapalot-chat)

**Reply by StephenDWright** [reply]
> Streamlit reruns the script from top to bottom for each interaction, so it's kind of tricky. I am searching for something similar. For now, I can come up with a lol solution for this 😆 :
> 
> ```python
> 
> # wait for the answer
> st.write("Bot is typing...")
> answer, source_documents = query_documents(user_input, selected_database, selected_collection)
> message(answer, key=answer_key, avatar_style="bottts", seed=5)
> ```
> I also used `with st.spinner("Processing..."):`
> 
> see: [my fork](https://github.com/sime2408/scrapalot-chat)

Thanks a lot for responding, will check out your fork. Yeah I realized there were issues but then I saw it being accomplished with gradio even though gradio has similar issues. I gather they are setting up some type of API to stream answers but not sure yet.

**Comment by zxjason** [comment]
I did a streamlit app for this

```
import streamlit as st
from dotenv import load_dotenv
from langchain.chains import RetrievalQA
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
from langchain.vectorstores import Chroma
from langchain.llms import GPT4All, LlamaCpp
import os
import time

load_dotenv()

embeddings_model_name = os.environ.get("EMBEDDINGS_MODEL_NAME")
persist_directory = os.environ.get('PERSIST_DIRECTORY')

model_type = os.environ.get('MODEL_TYPE')
model_path = os.environ.get('MODEL_PATH')
model_n_ctx = os.environ.get('MODEL_N_CTX')
model_n_batch = int(os.environ.get('MODEL_N_BATCH', 8))
target_source_chunks = int(os.environ.get('TARGET_SOURCE_CHUNKS', 4))

from constants import CHROMA_SETTINGS

def main():
    
    # Initialize the embeddings and retriever
    embeddings = HuggingFaceEmbeddings(model_name=embeddings_model_name)
    db = Chroma(persist_directory=persist_directory, embedding_function=embeddings, client_settings=CHROMA_SETTINGS)
    retriever = db.as_retriever(search_kwargs={"k": target_source_chunks})
    
    # Initialize the language model and QA system
    callbacks = [StreamingStdOutCallbackHandler()]
    llm = GPT4All(model=model_path, n_ctx=model_n_ctx, backend='gptj', n_batch=model_n_batch, callbacks=callbacks, verbose=False)
    qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=retriever, return_source_documents=True)

    # Set up Streamlit app
    st.title("Jason's PrivateGPT")
    
    if "answer" not in st.session_state:
        st.session_state.answer = ""
    if "docs" not in st.session_state:
        st.session_state.docs = []
    
    # Get user input
    query = st.text_input(f"**Enter a query:**")
    
    # Answer the question and display results
    if query:
        start = time.time()
        res = qa(query)
        answer, docs = res['result'], res['source_documents']
        end = time.time()

        st.session_state.answer = answer
        st.session_state.docs = docs

        st.write(f"**Question:** {query}")
        st.write(f"**Answer(Time {round(end - start, 2)} s):** {answer}")

        if docs:
            st.write(f"**Source:**")
            for document in docs:
                st.write(f"{document.metadata['source']}:")
                st.write(document.page_content)

if __name__ == "__main__":
    main()

```

**Reply by StephenDWright** [reply]
> I did a streamlit app for this
> 
> ```
> import streamlit as st
> from dotenv import load_dotenv
> from langchain.chains import RetrievalQA
> from langchain.embeddings import HuggingFaceEmbeddings
> from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
> from langchain.vectorstores import Chroma
> from langchain.llms import GPT4All, LlamaCpp
> import os
> import time
> 
> load_dotenv()
> 
> embeddings_model_name = os.environ.get("EMBEDDINGS_MODEL_NAME")
> persist_directory = os.environ.get('PERSIST_DIRECTORY')
> 
> model_type = os.environ.get('MODEL_TYPE')
> model_path = os.environ.get('MODEL_PATH')
> model_n_ctx = os.environ.get('MODEL_N_CTX')
> model_n_batch = int(os.environ.get('MODEL_N_BATCH', 8))
> target_source_chunks = int(os.environ.get('TARGET_SOURCE_CHUNKS', 4))
> 
> from constants import CHROMA_SETTINGS
> 
> def main():
>     
>     # Initialize the embeddings and retriever
>     embeddings = HuggingFaceEmbeddings(model_name=embeddings_model_name)
>     db = Chroma(persist_directory=persist_directory, embedding_function=embeddings, client_settings=CHROMA_SETTINGS)
>     retriever = db.as_retriever(search_kwargs={"k": target_source_chunks})
>     
>     # Initialize the language model and QA system
>     callbacks = [StreamingStdOutCallbackHandler()]
>     llm = GPT4All(model=model_path, n_ctx=model_n_ctx, backend='gptj', n_batch=model_n_batch, callbacks=callbacks, verbose=False)
>     qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=retriever, return_source_documents=True)
> 
>     # Set up Streamlit app
>     st.title("Jason's PrivateGPT")
>     
>     if "answer" not in st.session_state:
>         st.session_state.answer = ""
>     if "docs" not in st.session_state:
>         st.session_state.docs = []
>     
>     # Get user input
>     query = st.text_input(f"**Enter a query:**")
>     
>     # Answer the question and display results
>     if query:
>         start = time.time()
>         res = qa(query)
>         answer, docs = res['result'], res['source_documents']
>         end = time.time()
> 
>         st.session_state.answer = answer
>         st.session_state.docs = docs
> 
>         st.write(f"**Question:** {query}")
>         st.write(f"**Answer(Time {round(end - start, 2)} s):** {answer}")
> 
>         if docs:
>             st.write(f"**Source:**")
>             for document in docs:
>                 st.write(f"{document.metadata['source']}:")
>                 st.write(document.page_content)
> 
> if __name__ == "__main__":
>     main()
> 
> ```

This provides a streamlit interface but it doesn't address streaming responses as far as I can see.

**Reply by riyazmuthyalapati** [reply]
The UI doesn't work for me..

**Comment by Pedroman1** [comment]
So I have this thing running great on my computer at home but I want a way to access it from my phone (still doing the muscle work at home on my 4090) thru either a website or some other way. Is that what this gradio and streamit thing you all are talking about is ? I wish I could just post this on a private website to access it at work because other streaming services such as parsec for me to remotely access it are gone. 

So in other words... anyone have any idea to make privategpt public? I am not using it for confidental documents

**Reply by sime2408** [reply]
@Pedroman1 dm me in [discord of this repo](https://github.com/sime2408/scrapalot-chat), I'll explain what I did in google collab to host it on my example

**Comment by 77bgh** [comment]
Hello, I tried as well to create an UI, with 2 main q and keeping a history over the q&a , however I cannot make it to receive an answer over q about the documents I uploaded "Enter a question about the documents I uploaded" , any ideas please ? thx in advance

the "Enter a general question" is working and the answer provided is from llm .bin

![image](https://github.com/imartinez/privateGPT/assets/64399919/c551336b-127d-4c35-b9cb-bb4d8e80b5df)

**Reply by 77bgh** [reply]
from tkinter import *
from tkinter import filedialog
from tkinter.ttk import Progressbar
from threading import Thread
from functools import partial
import os
import shutil
import subprocess
import sys
import load
from tqdm import tqdm
from tkinter import messagebox
from time import sleep

directory_path = "your custom path here/privateGPT"

absolute_path = os.path.abspath(directory_path)

sys.path.append(absolute_path)
import runGPT

qa = runGPT.qa
parse_arguments = runGPT.parse_arguments

class ChatGUI:
    def __init__(self):
        self.window = Tk()
        self.window.title("ChatBot")
        self.create_widgets()
        self.question_history = []
        self.answer_history = []
        self.doc_question_history = [] 
        self.args = parse_arguments()
        self.result = None  
        self.menu = Menu(self.window)
        self.window.config(menu=self.menu)

        self.file_menu = Menu(self.menu, tearoff=False)
        self.file_menu.add_command(label="Upload and Index", command=self.upload_and_index)
        self.menu.add_cascade(label="File", menu=self.file_menu)

    def create_widgets(self):
        self.question_label = Label(self.window, text="Enter a general question:")
        self.question_label.pack()

        self.question_entry = Entry(self.window, width=50)
        self.question_entry.pack()

        self.doc_question_label = Label(self.window, text="Enter a question about the documents I uploaded:")
        self.doc_question_label.pack()

        self.doc_question_entry = Entry(self.window, width=50)
        self.doc_question_entry.pack()

        self.answer_label = Label(self.window, text="Answer:")
        self.answer_label.pack()

        self.answer_text = Text(self.window, height=10, width=50, wrap="word")
        self.answer_text.pack()

        self.submit_button = Button(self.window, text="Submit", command=self.submit_question)
        self.submit_button.pack()

        self.history_label = Label(self.window, text="Question History:")
        self.history_label.pack()

        self.history_text = Text(self.window, height=30, width=50, wrap="word")
        self.history_text.pack()

    def run_load_script(self, file_path):
        self.progress_window = Toplevel(self.window)
        self.progress_window.title("Progress")

        progress_label = Label(self.progress_window, text="Uploading and Indexing Files:")
        progress_label.pack()

        self.progress = Progressbar(self.progress_window, length=200, mode="determinate")
        self.progress.pack()

        self.total_files = len(file_paths)
        self.completed_files = 0

        with tqdm(total=self.total_files, ncols=60) as pbar:
            for file_path in file_paths:

                Thread(target=self.copy_file, args=(file_path,)).start()
                pbar.update(1)

                if "Split into" in output:
                    self.progress["value"] = 100
                    self.progress_window.after(1000, self.show_final_message)
                    break

                self.progress["value"] += 100 / self.total_files
                self.progress.update()        
                sleep(0.1)

    def show_final_message(self):
        self.progress_window.destroy()
        messagebox.showinfo("File Upload", "File uploaded and indexed successfully")

    def upload_and_index(self):
        file_paths = filedialog.askopenfilenames()
        if len(file_paths) > 0:
            Thread(target=partial(self.process_files, file_paths)).start()

    def copy_file(self, file_path):
        file_name = os.path.basename(file_path)
        destination_path = os.path.join("source_documents", file_name)
        shutil.copy(file_path, destination_path)

        output = subprocess.check_output(["python", "load.py", destination_path], universal_newlines=True)
        self.progress["value"] = 100

        if "No new valid documents to load" in output:
            messagebox.showinfo("Load.py Output", "No new valid documents to load")
        else:
            messagebox.showinfo("Load.py Output", output.strip())

        self.progress_window.after(1000, self.show_final_message)

    def process_files(self, file_paths):
        self.progress_window = Toplevel(self.window)
        self.progress_window.title("Progress")

        progress_label = Label(self.progress_window, text="Uploading and Indexing Files:")
        progress_label.pack()

        self.progress = Progressbar(self.progress_window, length=200, mode="determinate")
        self.progress.pack()

        self.total_files = len(file_paths)
        self.completed_files = 0

        with tqdm(total=self.total_files, ncols=60) as pbar:
            for file_path in file_paths:

                Thread(target=self.copy_file, args=(file_path,)).start()
                pbar.update(1)

    def close_progress_window(self):
        self.progress_window.destroy()

    def submit_question(self):
        question = self.question_entry.get()
        doc_question = self.doc_question_entry.get()
        self.question_entry.delete(0, END)
        self.doc_question_entry.delete(0, END)

        Thread(target=partial(self.get_answer, question, doc_question)).start()

    def get_answer(self, question, doc_question):
        if self.result is None:

            self.result = qa(question, doc_question, self.args)
        else:

            self.result = qa(question, doc_question, self.args)

        answer = self.result['answer']
        complete_answer = f"Question:\n{question}\n\nAnswer:\n{answer}"
        self.display_answer(complete_answer)
        self.update_history(question, doc_question, complete_answer)

    def display_answer(self, answer):
        self.answer_text.delete(1.0, END)
        self.answer_text.insert(END, answer)
        self.answer_text.tag_config("answer_tag", foreground="blue")
        self.answer_text.tag_add("answer_tag", "1.0", "end")
        
    def update_history(self, question, doc_question, answer):  
        self.question_history.append(question)
        self.doc_question_history.append(doc_question)  
        self.answer_history.append(answer)

        self.history_text.delete(1.0, END)
        for i in range(len(self.question_history)):
            q = self.question_history[i]
            dq = self.doc_question_history[i]  
            a = self.answer_history[i]
            self.history_text.insert(END, f"Question {i + 1}: ", "question_tag")
            self.history_text.insert(END, f"{q}\n", "question_tag")
            self.history_text.insert(END, f"Document Question {i + 1}: ", "question_tag")  
            self.history_text.insert(END, f"{dq}\n", "question_tag")  
            self.history_text.insert(END, f"Answer {i + 1}: ", "answer_tag")
            self.history_text.insert(END, f"{a}\n", "answer_tag")
            self.history_text.insert(END, "\n")

        self.history_text.tag_config("question_tag", foreground="red")
        self.history_text.tag_config("answer_tag", foreground="blue")

    def run(self):
        self.window.mainloop()
if __name__ == "__main__":
    chat_gui = ChatGUI()
    chat_gui.run()
